Alicerce/Sources/DeepLinking/Router.swift

```diff
@@ -0,0 +1,164 @@
+
+    enum Error: Swift.Error {
+        case invalidRoute(InvalidRouteError)
+        case duplicateRoute
+        case routeNotFound
+
+        enum InvalidRouteError: Swift.Error {
+            case misplacedEmptyComponent
+            case conflictingVariableComponent(existing: String, new: String)
+            case invalidVariableComponent(String)
+            case invalidURL
+            case unexpected(Swift.Error)
+        }
+    }
+
+    private var routes: [Route.Scheme : Tree] = [:]
+    private let queue: DispatchQueue
+
+    init(qos: DispatchQoS = .default) {
+        queue = DispatchQueue(label: "com.mindera.Alicerce.\(type(of: self)).queue", qos: qos)
+    }
+
+    func register(_ route: URL, handler: Handler) throws {
+        let (scheme, routeComponents) = parseAnnotatedRoute(route)
+
+        try queue.sync { [unowned self] in
+
+            do {
+                if var schemeTree = self.routes[scheme] {
+                    try schemeTree.add(route: routeComponents, handler: handler)
+                    self.routes[scheme] = schemeTree
+                } else {
+                    self.routes[scheme] = try Tree(route: routeComponents, handler: handler)
+                }
+            } catch Tree.Error.duplicateEmptyComponent {
+                throw Error.duplicateRoute 
+            } catch Tree.Error.invalidRoute {
+                 // empty route elements are only allowed at the end of the route
+                throw Error.invalidRoute(.misplacedEmptyComponent)
+            } catch let Tree.Error.conflictingParameterName(existing, new) {
+                throw Error.invalidRoute(.conflictingVariableComponent(existing: existing, new: new))
+            } catch {
+                throw Error.invalidRoute(.unexpected(error))
+            }
+        }
+    }
+
+    func unregister(_ route: URL) throws -> Handler {
+        let (scheme, routeComponents) = parseAnnotatedRoute(route)
+
+        return try queue.sync { [unowned self] in
+            guard var schemeTree = self.routes[scheme] else { throw Error.routeNotFound }
+
+            let handler: Handler
+
+            do {
+                handler = try schemeTree.remove(route: routeComponents)
+            } catch Tree.Error.routeNotFound {
+                throw Error.routeNotFound
+            } catch {
+                assertionFailure("🔥: unexpected Route.Tree.Error type!")
+                // FIXME: add logging
+                print("⚠️: unexpected error when unregistering \(route)! Error: \(error)")
+                throw Error.routeNotFound
+            }
+
+            self.routes[scheme] = schemeTree
+
+            return handler
+        }
+    }
+
+    func route(_ route: URL) throws {
+        let (scheme, routeComponents, queryItems) = try parseRoute(route)
+
+        let match: Match = try queue.sync { [unowned self] in
+            guard let schemeTree = self.routes[scheme] else { throw Error.routeNotFound }
+
+            let match: Match
```

Comment:
I completely agree. Don't know where I got the idea that I couldn't `return` directly from inside `do...catch` scopes 😅